File: day11.py
"""Solution to day 11 of Advent of Code"""

from get_input import get_input


def surround(location, seek):
    total = 0
    for diff in (1+1j, 1+0j, 1-1j, 0+1j, 0-1j, -1+1j, -1+0j, -1-1j):
        if seek(location, diff):
            total += 1
    return total


def update(grid, rules):
    new = grid.copy()
    for location in grid.keys():
        new[location] = rules(location)
    return new


def part1(seating):
    last = None
    while seating != last:
        last = seating

        def seek(pos, diff):
            return last.get(pos + diff) == '#'

        def rules(location):
            seat = last.get(location)
            if seat == "L" and surround(location, seek) == 0:
                return '#'
            elif seat == '#' and surround(location, seek) >= 4:
                return 'L'
            return seat

        seating = update(last, rules)
        assert len(last) == len(seating)
    return sum(1 for value in seating.values() if value == '#')


def part2(seating):
    last = None
    while seating != last:
        last = seating

        def seek(pos, diff):
            seat = last.get(pos+diff)
            if seat == '#':
                return True
            elif seat == 'L' or seat is None:
                return False
            return seek(pos+diff, diff)

        def rules(location):
            seat = last.get(location)
            if seat == "L" and surround(location, seek) == 0:
                return '#'
            elif seat == '#' and surround(location, seek) >= 5:
                return 'L'
            return seat

        seating = update(last, rules)
        assert len(last) == len(seating)
    return sum(1 for value in seating.values() if value == '#')


def parse(text):
    return {
        complex(r, c): seat
        for r, row in enumerate(text.strip().splitlines())
        for c, seat in enumerate(iter(row.strip()))
    }


TEST = """L.LL.LL.LL
LLLLLLL.LL
L.L.L..L..
LLLL.LL.LL
L.LL.LL.LL
L.LLLLL.LL
..L.L.....
LLLLLLLLLL
L.LLLLLL.L
L.LLLLL.LL"""


if __name__ == "__main__":
    assert part2(parse(TEST)) == 26
    LINES = parse(get_input(day=11, year=2020))
    print(f"Part 1: {part1(LINES)}")
    print(f"Part 2: {part2(LINES)}")

File: day11.rs
use common::get_input;
use common::point::Point;
use std::collections::HashMap;

type Input = HashMap<Point, char>;

fn surround(grid: &Input, location: &Point, seek: &dyn Fn(&Input, &Point, &Point) -> bool) -> usize {
    let mut total = 0;
    let points = vec![
        Point::new( 1, -1), Point::new( 1,  0), Point::new( 1,  1),
        Point::new( 0, -1),                     Point::new( 0,  1),
        Point::new(-1, -1), Point::new(-1,  0), Point::new(-1,  1),
    ];
    for diff in points {
        if seek(grid, &location, &diff) {
            total += 1;
        }
    }
    total
}

fn update(grid: &Input, rules: &dyn Fn(&Input, &Point) -> char) -> Input {
    let mut new = grid.clone();
    for (key, val) in new.iter_mut() {
        *val = rules(grid, key);
    }
    new
}

fn part1(seating: &Input) -> usize {
    let mut last = seating.clone();
    let mut current;
    loop {
        fn rules(grid: &Input, location: &Point) -> char {
            fn seek(grid: &Input, pos: &Point, diff: &Point) -> bool {
                grid.get(&(*pos + *diff)) == Some(&'#')
            }
            let seat = grid[location];
            if seat == 'L' && surround(&grid, location, &seek) == 0 {
                return '#'
            } else if seat == '#' && surround(&grid, location, &seek) >= 4 {
                return 'L'
            }
            seat
        };
        current = update(&last, &rules);
        if last == current {
            return current.values().filter(|&&c| c == '#').count()
        }
        last = current;
    }
}

fn part2(seating: &Input) -> usize {
    let mut last = seating.clone();
    let mut current;
    loop {
        fn rules(grid: &Input, location: &Point) -> char {
            fn seek(grid: &Input, pos: &Point, diff: &Point) -> bool {
                match grid.get(&(*pos + *diff)) {
                    Some('#') => true,
                    Some('L')|None => false,
                    Some('.') => seek(grid, &(*pos+*diff), diff),
                    _ => unimplemented!()
                }
            }
            let seat = grid[location];
            if seat == 'L' && surround(&grid, location, &seek) == 0 {
                return '#'
            } else if seat == '#' && surround(&grid, location, &seek) >= 5 {
                return 'L'
            }
            seat
        };
        current = update(&last, &rules);
        if last == current {
            return current.values().filter(|&&c| c == '#').count()
        }
        last = current;
    }
}

fn parse(text: String) -> Input {
    let mut seats = HashMap::new();
    for (r, row) in text.split('\n').enumerate() {
        for (c, seat) in row.chars().enumerate() {
            seats.insert(Point::new(r as i32, c as i32), seat);
        }
    }
    seats
}

fn main() {
    let input = parse(get_input(11, 2020));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
